stream: added join/leave notifications in private streams via notification bot.

s-bose7 opened the pull request:
user (un)subscription events now will be send out via notification bot in private streams.
[discussion.](https://chat.zulip.org/#narrow/stream/2-general/topic/Non.20PR.20code.20review)

<!-- Describe your pull request here.-->
Added two function in `zerver/actions/streams.py` and `zerver/views/streams.py`  to send user-left and user-join notifications respectively via notification bot. I also added few test cases to account for the new notifications.

Fixes: <!-- Issue link, or clear description.--> #2746

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:** As seen here there will be different message for adding 1, 2 or more users to a private stream, user-left messages are unchanged for every unsubscription.
![notification_bot](https://github.com/zulip/zulip/assets/69990740/023bc589-9703-4640-ad82-036ed93d3350)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

s-bose7 commented:
I've made certain changes in `zerver/actions/streams.py`, I'm pretty sure due to that change the queries for `send_events` got disrupted.

I've addressed a few of those, but still need to figure out what should be appropriate changes for rest of the failing test cases.
A quick review would definitely going to help me a lot.
Thanks!

alya commented:
I added a note about the text to use on the issue. We should be referring to users with silent mentions (see https://zulip.com/help/mention-a-user-or-group#silently-mention-a-user).

s-bose7 commented:
> I added a note about the text to use on the issue. We should be referring to users with silent mentions (see https://zulip.com/help/mention-a-user-or-group#silently-mention-a-user).

Okay, so should I use silent mentions to send subscribe/unsubscribe messages in a private stream by the notification bot?   
Do you think the current bot message for both events is okay? or should I refactor that too, I basically structure the bot message according to [this](https://github.com/zulip/zulip/pull/8652#discussion_r173963839). Let me know If this is looks appropriate.

alya commented:
You should use the message text/location I just added to the issue, not the text/location that were proposed in 2018. The broader context of the app has changed since then.

s-bose7 commented:
> You should use the message text/location I just added to the issue, not the text/location that were proposed in 2018. The broader context of the app has changed since then.

Understood! Thanks for notifying me about that. I will make sure to change it.

s-bose7 commented:
Hello @alya, I've made some refinements to my pull request. Kindly review it and share any suggestions for further improvement.

alya commented:
Thanks!

Please clean up your commit history, fix the failing tests, and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

s-bose7 commented:
Hello @alya, Thank you for your feedback. I've squashed my commits. I'm unable to run `./tools/test-api` on my local machine; for some reason, it's taking too long to start sending requests to the endpoints. As a result, I'm getting setbacks in addressing the failing API test on Debian 11.

If you look at the failing test case, you'll find that during sending a request to the endpoint `http://zulip.zulipdev.com:9981/api/v1/messages/43/reactions`, I'm getting an 'invalid message' as a response. The probable solution for that is to change the `'message_id'` to some appropriate integer value so that it matches whatever value there is in the test database for this endpoint. But I'm not sure about what would be the appropriate value. 

Your feedback would be highly appreciated!
Thanks! 

s-bose7 committed:
stream: add join/leave notifications in private streams.

constructing the data structures to hold information about the users who join/leave a private stream.
finally sending notifications via zulip notification bot in private streams upon triggering a subscribe/unsubscribe event.

added reasonable amount of test cases to account for new notifications from zulip notification bot.

fixes #2746

utkarsh-shrivastav77 commented:
@zulipbot claim

chandanSahoo-cs commented:
Can I be assigned this issue if it is still unsolved?

Sakshi-Singh2004 commented:
@zulipbot claim

cyrusgomes40 commented:
@zulipbot claim

AlphaGaurav13 commented:
@zulipbot claim

HeetRanpura commented:
@zulipbot I want to claim this issue as it has been staled for a longer period
